### PR TITLE
CBG-1386: Handle pruning of history

### PR DESCRIPTION
**Pull request by JRascagneres**
Adds pruning to grant history. Rules as follows:

 - If the history entries for a role / channel are not updated in 30 days then the entire history for that role / channel will be removed.
 - Grant Pairs are pruned to maintain a maximum of 10 entries and a minimum of 1 entry. Numbers in between are calculated in order to maintain grant history no larger than 1MB. 
 Note 1MB can be exceeded if an extreme number of channels / roles have history as each requires at least 1 entry.
 
The calculation for the second rule has a unit test.

Created AuthenticatorOptions to pass through options to the Authenticator. Currently has two attributes but allows for easy additions in future.

### CODE REVIEW

File: auth/auth_test.go [hunk 1]
```diff
@@ -2648,3 +2648,33 @@ func TestInvalidateChannels(t *testing.T) {
...
+		{
+			Name:                     "Tiny history length",
+			HistoryLengthInput:       1,
+			MaxEntriesExpectedOutput: 10,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.Name, func(t *testing.T) {
+			assert.Equal(t, CalculateMaxHistoryEntriesPerGrant(testCase.HistoryLengthInput), testCase.MaxEntriesExpectedOutput)
```
Comment (bbrks) [inline auth/auth_test.go]: expected <--> actual

```suggestion
			assert.Equal(t, testCase.MaxEntriesExpectedOutput, CalculateMaxHistoryEntriesPerGrant(testCase.HistoryLengthInput))
```

File: auth/auth_test.go [hunk 2]
```diff
@@ -2648,3 +2648,33 @@ func TestInvalidateChannels(t *testing.T) {
...
+		{
+			Name:                     "Extreme history length",
+			HistoryLengthInput:       1000000,
+			MaxEntriesExpectedOutput: 1,
+		},
+		{
+			Name:                     "Large history length",
+			HistoryLengthInput:       1000,
+			MaxEntriesExpectedOutput: 10,
+		},
```
Comment (bbrks) [inline auth/auth_test.go]: Include these so:
- We can see where the distribution is (these numbers are ballpark, they have a fair amount of leeway in them)
- The test fails if we change things to affect the distribution too significantly

```suggestion
		{
			Name:                     "Upper bound",
			HistoryLengthInput:       3600,
			MaxEntriesExpectedOutput: 1,
		},
		{
			Name:                     "Large history length",
			HistoryLengthInput:       3330,
			MaxEntriesExpectedOutput: 3,
		},
		{
			Name:                     "Median history length",
			HistoryLengthInput:       3075,
			MaxEntriesExpectedOutput: 5,
		},
		{
			Name:                     "Small history length",
			HistoryLengthInput:       2800,
			MaxEntriesExpectedOutput: 7,
		},
		{
			Name:                     "Lower bound",
			HistoryLengthInput:       2550,
			MaxEntriesExpectedOutput: 10,
		},
```

**Review by bbrks — CHANGES_REQUESTED** [review]

File: auth/role.go [hunk 3]
```diff
@@ -35,6 +36,19 @@ type roleImpl struct {
 
 type TimedSetHistory map[string]GrantHistory
 
+func (timedSet TimedSetHistory) PruneHistory(partitionWindow time.Duration) []string {
+	prunedChannelHistory := make([]string, 0)
+	for chanName, grantHistory := range timedSet {
+		grantTime := time.Unix(0, grantHistory.UpdatedAt)
```
[the remark this answers is not in the document]
Comment (JRascagneres) [reply auth/role.go]: Switched to Unix

File: base/constants.go [hunk 4]
```diff
@@ -167,6 +167,8 @@ var (
 	DefaultWarnThresholdChannelsPerDoc  = uint32(50)
 	DefaultWarnThresholdChannelsPerUser = uint32(50000)
 	DefaultWarnThresholdGrantsPerDoc    = uint32(50)
+	DefaultTimedSetHistoryPruneCap      = 10
```
[the remark this answers is not in the document]
Comment (JRascagneres) [reply base/constants.go]: Yeah woops, left this. 
I'll add a const for the max and min.

File: auth/auth.go [hunk 5]
```diff
@@ -222,20 +219,51 @@ func (auth *Authenticator) calculateHistory(invalSeq uint64, invalGrants ch.Time
...
+		currentHistoryForGrant.UpdatedAt = time.Now().Unix()
 		currentHistoryForGrant.Entries = append(currentHistoryForGrant.Entries, GrantHistorySequencePair{
 			StartSeq: previousInfo.Sequence,
 			EndSeq:   invalSeq,
 		})
 		currentHistory[previousName] = currentHistoryForGrant
 	}
 
+	if prunedHistory := currentHistory.PruneHistory(auth.ClientPartitionWindow); len(prunedHistory) > 0 {
+		base.Debugf(base.KeyCRUD, "rebuildChannels: Pruned principal history on %s for %s", base.UD(princName), base.UD(strings.Join(prunedHistory, ",")))
```
Comment (bbrks) [inline auth/auth.go]: Would be better to redact each element individually, so each channel name gets a hash, which is already supported for string slices.

```suggestion
		base.Debugf(base.KeyCRUD, "rebuildChannels: Pruned principal history on %s for %s", base.UD(princName), base.UD(prunedHistory))
```

**Review by bbrks — CHANGES_REQUESTED** [review]
Looks good apart from the last change for log redaction

**Review by bbrks — DISMISSED** [review]

File: auth/auth.go [hunk 6]
```diff
@@ -222,20 +218,51 @@ func (auth *Authenticator) calculateHistory(invalSeq uint64, invalGrants ch.Time
...
+			grantHistory.Entries[1].StartSeq = grantHistory.Entries[0].StartSeq
+			currentHistory[grantName] = grantHistory
+		}
+	}
+
 	return currentHistory
 }
 
+func CalculateMaxHistoryEntriesPerGrant(channelCount int) int {
+	const maximumHistoryBytes = 1024 * 1024 // 1MB
```
Comment (adamcfraser) [inline auth/auth.go]: All three of these constants are the same units, but that's not evident from naming.  What do you think about maximumHistoryBytes, averageHistoryKeyBytes, and averageHistoryEntryBytes?

**Review by adamcfraser — DISMISSED** [review]
Looks good, just some minor nits about naming.

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**